ext/neatdoc.py

```diff
@@ -0,0 +1,148 @@
+from sphinx.ext.autosummary import Autosummary
+from sphinx.ext.autosummary import get_documenter
+from sphinx import addnodes
+from docutils.parsers.rst import directives, Directive
+from docutils.nodes import Text
+from sphinx.util.inspect import safe_getattr,safe_getmembers
+import re
+import sphinx.ext.autodoc
+import sys
+import inspect
+
+class Docsummary(Autosummary):
+    has_content = True
+    required_arguments = 1
+    final_argument_whitespace = False
+    option_spec = {
+        'methods': directives.unchanged,
+        'attributes': directives.unchanged,
+        'modulecontent': directives.unchanged,
+        'inherited': directives.unchanged
+    }
+
+
+    @staticmethod
+    def get_members(obj, typ):
+        items = []
+        for name in dir(obj):
+            try:
+                member = safe_getattr(obj, name)
+                if type(member) is str or type(member) is list:
+                    documenter = None
+                else:
+                    documenter = get_documenter(member, obj)
+            except AttributeError:
+                continue
+            if documenter is not None:
+                if documenter.objtype == typ:
+                    items.append((name,member))
+
+        return items
+    @staticmethod
+    def get_mod_member(obj):
+        members = safe_getmembers(obj)
+        ret =[]
+        for mname in members:
+            try:
+                ret.append((mname, safe_getattr(obj, mname[0])))
+            except:
+                pass
+        return ret
+
+  
+    def run(self):
+        app = self.state.document.settings.env.app
+        self.content = []
+
+        if ('methods' in self.options) or ('attributes' in self.options):
+            clazz = self.arguments[0]
+            (module_name,class_name) = re.findall(re.compile("^(.*)\.([^.]*)$"),clazz)[0]
+            class_name = class_name#.encode("utf-8")
+            module_name = module_name#.encode("utf-8")
+            try:
+                m = __import__(module_name, globals(), locals(), [class_name])
+                c = getattr(m, class_name)
+            except ImportError:
+                print("Warning: Could not import %s's class %s" % (module_name,class_name))
+                return []
+
+            if 'methods' in self.options:
+                self.content_inherited = []
+                methods = self.get_members(c, 'method')
+
+                for method in methods:
+                    if method[0] not in c.__dict__:
+                        self.content_inherited.append("~%s.%s" % (clazz, method[0]))
+                    else:
+                        self.content.append("~%s.%s" % (clazz, method[0]))
+
+                if 'inherited' in self.options:
+                    self.content = self.content_inherited
+                    return super(Docsummary, self).run()
+                else:
+                    return super(Docsummary, self).run()
+
+
```

Comment:
so much space?

Reply:
Sorry! All extensions to Sphinx that are not listed in the `extensions` variable in `source/conf.py` are currently disabled. The folder `/ext` contains custom extensions, most of which are created by former MDP developers. I worked on creating neatdoc to make documentation creating easier and nicer, but it sort of became a dead end. I do not plan on working on it anymore.

I experienced errors with some of the custom extensions and the newer Sphinx versions - if needed I will describe these errors in more detail. It may be better to not rely on self-maintained custom extensions and convert the rst-code accordingly. Otherwise we should probably fix these and better document the extensions such that they be easier maintained in the future. I am not sure how long this would take.